firmware/library/L1_Drivers/pwm.hpp

```diff
@@ -1,13 +1,168 @@
+    {
+        kTimerMode = (3 << 0),
+        kCounterEnable = (1 << 0),
+        kResetMr0 = (1 << 1),
+        kPwmEnable = (1 << 3)
+    };
+    static LPC_PWM_TypeDef * pwm1;
+    static LPC_SC_TypeDef * sc;
+
+    template <unsigned pwm_channel>
+    static Pwm CreatePwm()
+    {
+        static_assert(pwm_channel <= 6,
+                      "Available PWM channels only go up to pin 6.");
+        return Pwm(pwm_channel);
+    }
+
+    explicit constexpr Pwm(uint8_t channel):
+        pwm_(pwm_pin_),
+        pwm_pin_(2, static_cast<uint8_t>(channel - 1)),
+        channel_(channel)
+    {
+    }
+
+    explicit constexpr Pwm(const PinConfigureInterface & pin):
+        pwm_(pin),
+        pwm_pin_(PinConfigure::CreateInactivePin()),
+        channel_(1)
+    {
+    }
+
+    ~Pwm()
+    {
+        pwm_pin_.SetPinFunction(kGpioFunction);
+    }
+
+    void Initialize(uint32_t frequency_hz = 1'000) override
+    {
+        SJ2_ASSERT_FATAL(channel_ < 6,
+         "Channel selected is out of bounds.");
+        // DIRECT MANIUPULATION OF SYSTEM CLOCK REGISTERS
+        // WILL REPLACE WITH FUNCTIONS LATER
+        // Enables PWM1 power/clock control bit
+        sc->PCONP |= SysClk::kActivatePwm1;
+        // TODO(delwin-lei): Remove when sysclock complete.
+        // END OF DIRECT MANIPULATION
+
+        // Resets PWMTC on Match with MR0
+        pwm1->MCR |= PwmConfigure::kResetMr0;
+        pwm1->MR0 = config::kSystemClockRate / frequency_hz;
+
+        // Enables PWM TC and PC for counting and enables PWM mode
+        EnablePwm();
+        pwm1->CTCR &= ~PwmConfigure::kTimerMode;
+
+        // Enables PWM[channel] output
+        pwm1->PCR |= PwmOutputEnable(channel_);
+
+        pwm_pin_.SetPinFunction(kPwmFunction);
+    }
+
+    void SetDutyCycle(float percentage) override
+    {
+        SJ2_ASSERT_FATAL(0.0 < percentage/100 && percentage/100 < 1.0,
+            "Percentage of Duty Cycle provided is out of bounds.");
+        EnablePwm(false);
+        *match_register_table[channel_] = CalculateDutyCycle(percentage);
+        pwm1->LER |= (1 << channel_);
+        EnablePwm();
+    }
+
+    float GetDutyCycle() override
+    {
+        return (static_cast<float>(*match_register_table[channel_]) /
+            static_cast<float>(GetMatchRegister0())) * 100;
+    }
+
+    void SetFrequency(uint32_t frequency_hz) override
+    {
+        // Disables PWM mode; this will reset all counters to 0
+        // And allow us to update MR0
+        EnablePwm(false);
+        pwm1->MR0 = config::kSystemClockRate / frequency_hz;
+        EnablePwm();
+    }
+
+    uint32_t GetFrequency() override
+    {
+        return config::kSystemClockRate / GetMatchRegister0();
+    }
+
+    void EnablePwm(bool enable = true) override
+    {
+        if(enable)
+        {
+            pwm1->TCR |= PwmConfigure::kCounterEnable | PwmConfigure::kPwmEnable;
+        }
+        else
+        {
+            pwm1->TCR &= ~PwmConfigure::kPwmEnable;
```

Comment:
Why don't you clear the counter enable as well? Or is it not necessary if you simply disable the whole pwm module.